libraries/Microsoft.Bot.Connector/Authentication/MicrosoftAppCredentials.cs

```diff
@@ -63,28 +60,54 @@ public MicrosoftAppCredentials(string appId, string password)
         public virtual string OAuthEndpoint { get { return AuthenticationConstants.ToChannelFromBotLoginUrl; } }
         public virtual string OAuthScope { get { return AuthenticationConstants.ToChannelFromBotOAuthScope; } }
 
-        protected readonly string TokenCacheKey;
 
         /// <summary>
-        /// Adds the host of service url to <see cref="MicrosoftAppCredentials"/> trusted hosts.
+        /// TimeWindow which controlls how often the token will be automatically updated
         /// </summary>
-        /// <param name="serviceUrl">The service URL.</param>
-        /// <remarks>If expiration time is not provided, the expiration time will DateTime.UtcNow.AddDays(1).</remarks>
-        public static void TrustServiceUrl(string serviceUrl)
-        {
-            TrustServiceUrl(serviceUrl, DateTime.UtcNow.Add(TimeSpan.FromDays(1)));
-        }
+        public static TimeSpan AutoTokenRefreshTimeSpan { get; set; } = TimeSpan.FromMinutes(10);
 
         /// <summary>
         /// Adds the host of service url to <see cref="MicrosoftAppCredentials"/> trusted hosts.
         /// </summary>
-        /// <param name="serviceUrl">The service URL.</param>
-        /// <param name="expirationTime">The expiration time after which this service url is not trusted anymore.</param>
-        public static void TrustServiceUrl(string serviceUrl, DateTime expirationTime)
+        /// <param name="serviceUrl">The service url</param>
+        /// <param name="expirationTime">The expiration time after which this service url is not trusted anymore</param>
+        /// <remarks>If expiration time is not provided, the expiration time will DateTime.UtcNow.AddDays(1).</remarks>
+        public static void TrustServiceUrl(string serviceUrl, DateTime expirationTime = default(DateTime))
         {
-            lock (_trustedHostNamesSync)
+            try
             {
-                _trustedHostNames[new Uri(serviceUrl).Host] = expirationTime;
+                if (expirationTime == default(DateTime))
+                {
+                    // by default the service url is valid for one day
+                    var extensionPeriod = TimeSpan.FromDays(1);
```

Comment:
Most of this code should be removed. 

If the user passes in a time, we should use that time. I "fixed" this in the older code, and it looks like it's come back here. The rules this code is implementing are very bug prone, fragile, and will lead to issues. They should be removed.

Reply:
kept your trust url code

---
In reply to: [198676297](https://github.com/Microsoft/botbuilder-dotnet/pull/702#discussion_r198676297) [](ancestors = 198676297)